File: OrderChainQueue/C++/OrderQueue.cpp
#include "OrderQueue.h"

OrderQueue::OrderQueue()
{
	m_queue.front = 0;
	m_queue.rear = 0;
}

bool OrderQueue::Push(int nValue)
{
	if ((m_queue.rear + 1)% QUEUE_SIZE == m_queue.front) // 队列已满
		return false;

	m_queue.data[m_queue.rear] = nValue;
	m_queue.rear = (m_queue.rear + 1) % QUEUE_SIZE;
	return true;
}

bool OrderQueue::Pop(int* nValue)
{
	if (m_queue.rear == m_queue.front) // 队列为空
		return false;

	*nValue = m_queue.data[m_queue.front];
	m_queue.front = (m_queue.front + 1) % QUEUE_SIZE;
	return true;
}

void OrderQueue::Clear()
{
	m_queue.rear = 0;
	m_queue.front = 0;
}

File: OrderChainQueue/Go/order_queue.go
package main

//顺序队列
type OrderQueue struct {
	front int
	rear  int
	data  [MaxSize]int
}

func CreateOrderQueue() *OrderQueue {
	Q := &OrderQueue{
		front: 0,
		rear:  0,
	}
	return Q
}

// 入队
func (this *OrderQueue) Push(nValue int) bool {
	// 队列已满
	/*
	* 取模:
	* 1. 小于 目标值, 结果为本身
	* 2. 等于 目标值, 结果为0
	* 3. 大于 目标值, 结果为余数
	 */
	if (this.rear+1)%len(this.data) == this.front {
		return false
	}
	//入队
	this.data[this.rear] = nValue
	// 尾指针后移,小于长度取自身+1
	this.rear = (this.rear + 1) % len(this.data)
	return true
}

// 出队
func (this *OrderQueue) Pop(nValue *int) bool {
	//队列为空
	if this.rear == this.front {
		return false
	}
	*nValue = this.data[this.front]
	// 头指针后移
	this.front = (this.front + 1) % len(this.data)
	return true
}

// 清空队列
func (this *OrderQueue) Clear() {
	//
	this.front = 0
	this.rear = 0
}
